### ISSUE TITLE
Filename starting with dash fail with most hooks

### ISSUE THREAD

**Issue by dzamlo** [cachix/pre-commit-hooks.nix#303]
For most hooks, filenames are passed as argument to the command. A filename that start with a dash is then interpreted by the command as an option instead of a filename.

The standard way to avoid that is to pass the `--` argument to the command before the filenames. But of course this is not supported by all tools and need to be tested with each one.

## Steps to reproduce
Enable the `shfmt` hook.
Create a file named `--foobar.sh`
Add this file to git: `git add -- --foobar.sh`
Try to commit

**Comment by dzamlo** [comment]
> I believe this is more of an issue with pre-commit itself, no?

The pre-commit hooks commands, that should be fixed, are defined in this repo (the `entry` variables in `modules/hooks.nix`).

> Regardless, I would strongly discourage you from prefixing your files with --!

I agree that this is a bad idea, but I also think that tools should not break if this happens.

**Comment by sandydoo** [comment]
> The pre-commit hooks commands, that should be fixed, are defined in this repo (the entry variables in modules/hooks.nix).

Sure, I see what you're saying. It seems to me like the wrong place to deal with this. Does every hook author need to be aware of this and remember to end their commands with `--`?

Another approach is to prefix every path with `./`, which is something that pre-commit could do.

I set up a test repo with vanilla `pre-commit` and two hooks: [`trailing-whitespace`](https://github.com/pre-commit/pre-commit-hooks) and [`shfmt`](https://github.com/scop/pre-commit-shfmt). They both failed to run on files prefixed with `--`.

```yaml
repos:
  - repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v1.2.3
    hooks:
      - id: trailing-whitespace
  - repo: https://github.com/scop/pre-commit-shfmt
    rev: v3.6.0-2
    hooks:
      - id: shfmt
```
```shell
$ pre-commit run --all-files
Trim Trailing Whitespace.................................................Failed
- hook id: trailing-whitespace
- exit code: 2

usage: trailing-whitespace-fixer
       [-h]
       [--no-markdown-linebreak-ext]
       [--markdown-linebreak-ext [*|EXT[,EXT,...]]]
       [filenames ...]
trailing-whitespace-fixer: error: unrecognized arguments: --shell.sh --test.py

shfmt....................................................................Failed
- hook id: shfmt
- exit code: 2

flag provided but not defined: -shell.sh
usage: shfmt [flags] [path ...]
```

**Comment by dzamlo** [comment]
I agree, I opened an issue on the pre-commit repo: https://github.com/pre-commit/pre-commit/issues/2896

**Comment by LenaWil** [comment]
> Does every hook author need to be aware of this and remember to end their commands with `--`?

Just my two cents and unasked opinion and I will admit I'm not using this tool and got linked this issue through pre-commit, but yes!

So often a tool breaks because I dared to use a space or some other special character and because the build process is often build of multiple layers it's often quite hard to find out which tool was responsible. And just using `--` to fix it seems worth it to me to fix it.

**Comment by sandydoo** [comment]
> Does every hook author need to be aware of this and remember to end their commands with `--`?
> ...yes! ...And just using `--` to fix it seems worth it to me to fix it.

To clarify, I'm _not_ against fixing this! I'm trying to illustrate the two vastly different approaches to doing so.

One approach is to go through [every single hook on this list](https://pre-commit.com/hooks.html) (235 ! repos), test and fix the command, and then plead your case to the maintainers.  You've now involved hundreds of maintainers and spent thousands of person-hours.

Or you tackle the issue at the source with 1 PR. And as [this comment admits](https://github.com/pre-commit/pre-commit/issues/2896#issuecomment-1559518739), pre-commit is in charge of calling the tools and passing in the file paths. It's truly a shame that the linked [pre-commit issue](https://github.com/pre-commit/pre-commit/issues/2896) is now locked, because there's still a case to be argued there.
